JS backend post-merge update (#17)
input-output-hk/engineering: blog/2022-12-05-ghc-js-backend-merged.md

---
slug: 2022-12-05-ghc-js-backend-merged
title: JavaScript backend merged into GHC

> authors: [sylvain, doyougnu]

Reviewer: Did Luite not partake in this?

Reviewer 2: Not that I'm aware of.

Author: And Moritz!

tags: [ghc, javascript, cross-compilation]
---

For most of the year 2022, we, the GHC DevX team at IOG, have been working on
implementing a JavaScript backend for GHC, based on GHCJS. It has just been
[merged](https://gitlab.haskell.org/ghc/ghc/-/commit/cc25d52e0f65d54c052908c7d91d5946342ab88a)
into the master branch on November 30th, 2022. This post should answer most of

> the questions you may have about this backend.

Reviewer 2: resolved in ba2541fba4f03e62ca4f1b23b16090b8026246d9

[Skip to build instructions for the impatient](#build)

## Why Javascript?

> The only reason to target Javascript is that it is (sadly) the most portable

Reviewer: Let's not say sadly. It is *one* of the most portable languages due to a proliferation of JS runtimes (JavaScript core, v8, duck, quickjs, ???).

Reviewer 2: resolved in https://github.com/input-output-hk/engineering/commit/7329d1d36a5b0824e3ac812dd35bfc747d2048f0

language especially to write user interfaces.

> What about WebAssembly? WebAssembly is a more promising target language for

Reviewer: Webassembly is a promising target as well, and tweag has just merged a WASM backend into GHC as well. Webassembly is not as ubiquitous as JavaScript yet, and has a harder time interacting with JavaScript directly. Hence we believe that both backends provide different strengths and it is to our benefit to back both code generation paths in GHC for different use cases and requirements.

Reviewer 2: resolved in 7329d1d36a5b0824e3ac812dd35bfc747d2048f0

@angerman I used your words directly. In my book that means you should be a co-author. Is that ok with you? (Perhaps this is my latent academic side coming through :P)

of using Haskell instead is code sharing: frontend is Haskell code compiled to
Javascript and backend is Haskell code compiled to machine code. In particular
the (de)serialization code (e.g. from/to JSON) is shared and cannot get out of
sync between the frontend and the backend.

A lot of IOG code is written in Haskell. Duplicating some of this code into some
other language(s) would be a maintenance burden. That's why we invested into
this Haskell to Javascript compiler.

## Why a GHC backend?

Haskell is a language driven by its implementation in GHC. GHC development is
very active and the GHC API is totally unstable. As a consequence, maintaining a
Haskell to Javascript compiler based on GHC but outside of GHC is costly.

The maintenance burden is not hypothetical; our teammate Luite Stegeman has been
developing GHCJS for close to 10 years and has experienced the pain first hand.
GHCJS relied on a modified fork of the GHC library to suit GHCJS's needs. So any
changes to upstream GHC had to be adapted to the customized fork or GHCJS would
fall behind. And fall behind it did, as of writing GHCJS has stuck to using GHC
8.10; lagging behind by three major releases and counting.

> To compounding the issue, the normal Haskell toolchain does was not designed for

Reviewer 3: ```suggestion
To compoundi the issue, the normal Haskell toolchain was not designed for
```

Reviewer 2: fixed it 0290deef6ef1d53d17d39e78d71b97a20b5cb7f9 thanks!

an edge case like GHCJS. So GHCJS required that the normal tooling, e.g., Cabal
and Stack, could distinguish between upstream GHC library code and GHCJS code.
This meant that the GHCJS developers had to maintain the GHC fork, develop
GHCJS, and patch or contribute to Cabal and Stack. Simply put, the maintenance
burden was much too high per developer.

So instead of spending engineering time and energy _responding_ to ecosystem
changes and maintenance, the DevX team decided the best course of action was
enhancing GHC's cross-compilation support and add a proper JS backend based on
GHCJS. We feel that this adds value to the entire Haskell ecosystem, keeps the
JS backend in sync with GHC, provides a better user experience for all, and

> reduces maintenance costs.

Reviewer 2: resolved in 0e4e68e14086b420efdf747099759d24eadcfa55

## Is GHCJS dead?

Not yet! As it stands, the JS backend doesn't provide all the features provided
by GHCJS. In particular it doesn't support Template Haskell and we've removed
the GHCJS FFI to refine its design. See our roadmap below for more details.

Nevertheless GHCJS is unlikely to be updated to use a GHC version more recent
than 8.10.x. So from our point of view it is in maintenance mode until the JS
backend totally subsumes its features. However it's an open source project so
feel free to offer patches to update it to use newer version of the GHC library.

## What is missing from GHCJS?

The JS backend borrows a lot of code from GHCJS, but not all of it. If you are a
GHCJS user, here are the main differences:

> 1. GHCJS was stuck to GHC 8.10 while the JS backend follows GHC head.

Reviewer 3: There are two points number one :  1, 1, 2, 3

Reviewer 2: nice catch! This also made it into 0290deef6ef1d53d17d39e78d71b97a20b5cb7f9

   the JS backend in a principled way (e.g. by first getting evidences with
   ticky profiles, etc.).

## What's on the JS backend roadmap?

Our top priorities are:

- Implementing Template Haskell support
- Reducing generated JS code size
- Modernizing the generated JS code
- Enhancing runtime performance

## What has improved compared to GHCJS?

Or, why did it take you so long to port a stripped GHCJS into GHC? While it may
seem like such a task should be relatively quick&#151especially in a language
with such a good refactoring story like haskell&#151there were numerous road
blocks that we needed to remove before adding the backend. In particular, here
were the troublesome bits:

#### Removing the use of external libraries

GHCJS used libraries that aren't already dependencies of GHC: text, lens,
megaparsec, aeson... As we didn't want to add new dependencies to GHC, we've
refactored the code to avoid them. Examples:

- we've replaced Text with GHC's ShortText (which provides a similar API) and
  finally with GHC's FastString in most cases (which is usually more performant)
- we've replaced a lot of code using lenses with its understandable

>   counterpart (which is more common in GHC)

Reviewer 2: resolved in 0c3951ea99a52ea08bbdd781e63216e821f7a229

I hope that isn't true *cries in compile times*

ghcjs-prim. We've merged those into the existing base and ghc-prim libraries.

#### Reusing GHC's build system: Hadrian

GHCJS is known to be complex to build, relying on custom build scripts to deal
with the GHC fork it uses, etc. The JS backend however is as easy to build as
any other GHC. It doesn't require any wrapper script, only the "emconfigure"
tool provided by the Emscripten project.

You can now build a GHC with the JS backend with just these commands:

> ./boot
> emconfigure ./configure --target=js-unknown-ghcjs
> ./hadrian/build --bignum=native -j

Note that if this doesn't work, up to date instructions and troubleshootings can
be found on https://gitlab.haskell.org/ghc/ghc/-/wikis/javascript-backend

The Hadrian build system has been adapted to support Cabal's `js-sources`
stanzas that are to support user-provided `.js` files. Both the `rts` and `base`
packages required this feature.

#### Support for running GHC's testsuite

We can now run GHC's testsuite with the JS backend enabled! We had to tweak
Hadrian to make this possible (support for cross-compilers is subpar), but the
testsuite already found some bugs that we have since fixed.

> However, in order to merge for the GHC 9.6 release we had to be disable many

Reviewer 3: ```suggestion
However, in order to merge for the GHC 9.6 release we had to  disable many
```

isn't modular enough to support two environments at once: one for the target
code (JS code here) and one for the host (e.g. native x86 or AArch64 code for
the plugin). We've spent a lot of time making it more modular (see the
[Modularizing GHC](https://hsyl20.fr/home/files/papers/2022-ghc-modularity.pdf)
white paper we published earlier this year and Sylvain's [lightning
talk](https://youtu.be/OHGH5HLOCEM) at HIW 2022) but there is a lot more to do
to achieve this (cf
[#17957](https://gitlab.haskell.org/ghc/ghc/-/issues/17957)).

GHCJS used a fragile hack to support plugins: at plugin loading time it would
substitute the plugin unit with another corresponding one from another
package database.
It was fragile because it could violate GHC's single environment assumptions.

GHCJS's hack did not get ported. Nevertheless we have implemented a new way for
GHC to load plugins directly from libraries instead of packages
([#20964](https://gitlab.haskell.org/ghc/ghc/-/issues/20964)/[!7377](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7377)).
This method doesn't require GHC to load module interfaces for the plugin and its
dependencies, hence workarounds GHC's limitations.

> ## What about libraries using C sources?

Reviewer 2: I think its great to mention this, but should not be more than a sentence or two. A sign post for that blogpost perhaps. Thoughts @hsyl20 ?

> ## What about libraries using C sources?

Author: Yeah I would prefer that we describe this approach in another blog post. We will have to learn how you did it and to try it with the JS backend. It will take some time and I would prefer not to delay this post for too long.

provide shims so these JS sources will have to be upstreamed or reimplemented in
Haskell.

Note that the linking behavior is different due to the interpreted nature of
Javascript. In the JS backend, we can link with libraries using foreign imports
_even if_ the imported functions don't exist. Instead of failing at link time
(what usually happens with native code) a JS exception be raised only when and
if the imported function is called.

## How to help?

We have now reached our first milestone; anyone can easily build and test the JS
backend, and anyone can open bug reports or offer patches for the JS backend on
GHC's gitlab.

For those who offered their help this year: thank you! Until now it was
difficult to split the work into independent tasks (one fix led to a new
failure, which lead to an architectural issue etc.) and it was difficult to
coordinate with people outside of our team. However, we're now in a much better
position to discuss suggestions and to test/review patches in the spirit of open
source.

## tl;dr Just tell me how to say hello world {#build}

You need:

  - [emscripten](https://emscripten.org/docs/getting_started/downloads.html) version 3.14 or better

>   - llvm 15, or a patched llvm 

Author: I'm not sure about this: LLVM is bundled with emscripten usually, no?

Reviewer 4: Yeah we only need emscripten which comes with the whole toolchain, the llvm version should be implied by the emscripten version requirements.

(unless we now depend directly on some llvm tool in an update that I missed)

---
slug: 2022-12-05-ghc-js-backend-merged
title: JavaScript backend merged into GHC
authors: [sylvain, doyougnu, luite, josh, moritz]
tags: [ghc, javascript, cross-compilation]
---

A new Javascript backend has been
[merged](https://gitlab.haskell.org/ghc/ghc/-/commit/cc25d52e0f65d54c052908c7d91d5946342ab88a)

> into GHC on November 30th, 2022!

Author: I've used your sentence as-is, I hope it's ok.

I thought about mentioning that it's not any GHC that will be able to produce JavaScript code, but only a GHC built for the JavaScript target platform, but it's quite technical.

I wonder if we should tame expectations right in the introduction to avoid disappointment: not all of Haskell is supported (cf TH), it's not that easy to install a GHC with the JavaScript backend enabled, most of the libraries using C codes won't work, performance and code sizes aren't great, etc. Shouldn't we indicate that it's a feature preview or something like this? Even/especially for non techies it may be useful to know that JS support is coming but that it's not fully ready yet.

In this post, we, the GHC DevX team at IOG, describe the challenges we faced
bringing GHCJS to GHC, how we overcame those challenges, and what's left to do.
We hope this post will answer most of the questions you may have about the
backend and look forward to the community response.

[Skip to build instructions for the impatient](#build)

## Why Javascript?

We chose to target JavaScript because JavaScript is simply the most portable

> language on the web, and especially so for user interfaces. In addition to

Reviewer 5: The UI claim at the front would be better backed-up by a hello world running in a call to `alert` in the generated JS code. But I guess that without an FFI, that will be hard to do.

Reviewer 4: There is FFI, just not the advanced syntax that GHCJS had. Perhaps it needs to be clarified.

portability, there is a large amount of mature tooling around the JavaScript
language and such a large user base that the language will likely never die.

Webassembly is a promising target as well, and [Tweag](https://www.tweag.io/)
has just merged a

> [WASM backend](https://www.tweag.io/blog/2022-11-22-wasm-backend-merged-in-ghc/)

Reviewer 5: ```suggestion
[WebAssembly backend](https://www.tweag.io/blog/2022-11-22-wasm-backend-merged-in-ghc/)
```

our benefit, and the communities' to back both code generation paths in GHC for
different use cases and requirements.

## Why Haskell?

Other languages such as [PureScript](https://www.purescript.org/) target
Javascript and provide a programmer experience close to Haskell's. The benefit
of using Haskell instead is code sharing: frontend is Haskell code compiled to
Javascript and backend is Haskell code compiled to machine code. In particular
the (de)serialization code (e.g. from/to JSON) is shared and cannot get out of
sync between the frontend and the backend.

A lot of IOG code is written in Haskell. Duplicating some of this code into some
other language(s) would be a maintenance burden. That's why we invested into
this Haskell to Javascript compiler.

## Why a GHC backend?

Haskell is a language driven by its implementation in GHC. GHC development is
very active and the GHC API is totally unstable. As a consequence, maintaining a
Haskell to Javascript compiler based on GHC but outside of GHC is costly.

The maintenance burden is not hypothetical; our teammate Luite Stegeman has been
developing GHCJS for close to 10 years and has experienced the pain first hand.

> GHCJS relied on a modified fork of the GHC library to suit GHCJS's needs. So any

Reviewer 4: Perhaps the technical details don't matter so much.

GHCJS wasn't a fork of GHC, but a compiler that used the `ghc` library. Due to some limitations in the `ghc` library it couldn't use the library that came the already installed GHC, and the library had to be forked. Eventually the entire `ghc` library was pulled into the `ghcjs` library used by the GHCJS programs, with a few modifications.

Perhaps calling GHCJS "a modified GHC" rather than "fork" is enough detail to get the point across.

changes to upstream GHC had to be adapted to the customized fork or GHCJS would
fall behind. And fall behind it did, as of writing GHCJS has stuck to using GHC
8.10; lagging behind by three major releases and counting.

> 

Author: Add mention of Eta project here as suggested by @angerman

Author: Done

To compound the issue, the normal Haskell toolchain was not designed for an

> edge case like GHCJS. So GHCJS required that the normal tooling, e.g., Cabal and

Reviewer 4: If this is unclear, we should mention the different tool names (`ghc-pkg` -> `ghcjs-pkg`, `js_o`/`js_hi`/`js_a` extensions) and the versioning scheme where GHCJS had two versions, the GHCJS revision and the GHC version it was based on.

> edge case like GHCJS. So GHCJS required that the normal tooling, e.g., Cabal and

Reviewer 5: I would absolutely not mention all the different tool versions! Most users of GHC have only a vague idea that `ghc-pkg` exists, and learning about it here seems unlikely to teach them what you want to. What about a more conceptual explanation of why the JS backend is different than the native ones, aside from cross-compilation?

> edge case like GHCJS. So GHCJS required that the normal tooling, e.g., Cabal and

Reviewer 4: Hmm, hard to find a good balance between conciseness and clarity.

We can avoid going into details and say that GHCJS used different filenames and formats (`.js_o` extension for object files and `library.js_a` archives that aren't standard `ar` files if we need an example, footnote?), and that we've switched to the standard formats/names now.

Another non-standard thing was the GHCJS version (as a result of being a separate program using the `ghc` library), for example `GHCJS-0.1.1-8.6.3` for a GHCJS version 0.1.1 based on the `ghc-8.6.3` library. Cabal had to support that. Integrating into GHC lets us drop the double version.

Perhaps we can summarize it as: _Having non-standard file names, file formats and compiler version numbering._

I feel more detail is probably a bit out of scope for this post. The reason for the "differentness" is often historical

> edge case like GHCJS. So GHCJS required that the normal tooling, e.g., Cabal and

Author: I've added some examples of differences between GHCJS and GHC.

> Stack, could distinguish between upstream GHC library code and GHCJS code. This

Author: I think it was meant to be "compiler" instead of "library code". I've changed it this way.

meant that the GHCJS developers had to maintain the GHC fork, develop GHCJS, and
patch or contribute to Cabal and Stack. Simply put, the maintenance burden was
much too high per developer.

So instead of spending engineering time and energy _responding_ to ecosystem
changes and maintenance, the DevX team decided the best course of action was
enhancing GHC's cross-compilation support and add a proper JS backend based on
GHCJS. We feel that this adds value to the entire Haskell ecosystem, keeps the
JS backend in sync with GHC, provides a better user experience for all, reduces
maintenance costs, and greatly improves the backends in GHC in general. By
implementing support for a JS backend in GHC, we also improve GHC's support for
cross-compilation (and testing cross-compilers), which is directly applicable to
the WASM, iOS, and Android backends in GHC.

## Is GHCJS dead?

Not yet! As it stands, the JS backend doesn't provide all the features provided
by GHCJS. In particular it doesn't support Template Haskell and we've removed

> the GHCJS FFI to refine its design. See our roadmap below for more details.

Reviewer 4: remove the extended GHCJS FFI syntax, not all FFI

Nevertheless GHCJS is unlikely to be updated to use a GHC version more recent
than 8.10.x. So from our point of view it is in maintenance mode until the JS
backend totally subsumes its features. However it's an open source project so
feel free to offer patches to update it to use newer version of the GHC library.

## What is missing from GHCJS?

The JS backend borrows a lot of code from GHCJS, but not all of it. If you are a
GHCJS user, here are the main differences:

> 1. GHCJS was stuck to GHC 8.10 while the JS backend follows GHC head.

Reviewer 2: resolved in d9d7100383e5768e3683bf8a8b7c363afc32722e

2. GHCJS's incremental linking support ("base" bundles) hasn't been ported. This
   feature required too many changes (such as adding new command-line flags) and
   would have been backend-specific. This might be implemented in the future if
   it proves to be useful for the newer TH implementation for example.

3. GHCJS's JS code optimizer hasn't been ported. The code was trying to do too
   much all at once and consequently was fragile and slow. We plan to work on an
   intermediate representation between STG and JS to perform the same
   optimizations in a (type) safer/faster way.

4. GHCJS's compactor (link time optimizations) code hasn't been ported. Some
   optimizations have been reimplemented (e.g. global renaming of local
   identifiers), but some other are lacking (e.g. compacting initialization code).
   We plan to work on this as part of a larger effort on refactoring the code
   generator, the linker, and some runtime system aspects.
   See also [#22352](https://gitlab.haskell.org/ghc/ghc/-/issues/22352)

5. GHCJS's support for plugins hasn't been ported. The implementation was

>    unsafe (exchanging a unit for another when GHC tries to load a plugin).

Author: Hmm given that even SPJ doesn't know it, I guess not many.

Since Backpack, it's the term used to refer to 'a component of a Cabal package that has been compiled and (usually) installed in a "package database", which is really a "unit database"'.

I've removed the technical details and the comment about GHC's lack of modularity.

8. Any command-line flag introduced by GHCJS has not been ported. We didn't make
   any change to GHC's command-line in this work except for adding a `-ddump-js`
   flag. Other options will be added later as needed.

9. The JS backend itself hasn't been optimized and we even removed some
   (seemingly random) uses of NFData from GHCJS's code. We intend to optimize
   the JS backend in a principled way (e.g. by first getting evidences with
   ticky profiles, etc.).

## What's on the JS backend roadmap?

Our top priorities are:

- Implementing Template Haskell support
- Reducing generated JS code size
- Modernizing the generated JS code
- Enhancing runtime performance

## What has improved compared to GHCJS?

Or, why did it take you so long to port a stripped GHCJS into GHC? While it may
seem like such a task should be relatively quick&#151especially in a language
with such a good refactoring story like haskell&#151there were numerous road
blocks that we needed to remove before adding the backend. In particular, here
were the troublesome bits:

#### Removing the use of external libraries

GHCJS used libraries that aren't already dependencies of GHC: text, lens,

> megaparsec, aeson... As we didn't want to add new dependencies to GHC, we've

Reviewer 4: `attoparsec` I think, not megaparsec?

#### Reusing GHC's build system: Hadrian

GHCJS is known to be complex to build, relying on custom build scripts to deal
with the GHC fork it uses, etc. The JS backend however is as easy to build as
any other GHC. It doesn't require any wrapper script, only the "emconfigure"
tool provided by the Emscripten project.

You can now build a GHC with the JS backend with just these commands:

> ./boot
> emconfigure ./configure --target=js-unknown-ghcjs
> ./hadrian/build --bignum=native -j

Note that if this doesn't work, up to date instructions and troubleshootings can
be found on https://gitlab.haskell.org/ghc/ghc/-/wikis/javascript-backend

The Hadrian build system has been adapted to support Cabal's `js-sources`
stanzas that are to support user-provided `.js` files. Both the `rts` and `base`
packages required this feature.

#### Support for running GHC's testsuite

We can now run GHC's testsuite with the JS backend enabled! We had to tweak
Hadrian to make this possible (support for cross-compilers is subpar), but the
testsuite already found some bugs that we have since fixed.

However, in order to merge for the GHC 9.6 release we had to disable many tests

> because of missing features (TH, HPC, compact regions, etc.) or because the

Author: Indeed, done.

generated code would timeout (not surprising given the missing optimizer and
compactor).

But in the process of disabling those tests we've laid a good path forward.
We've added more precise properties to the testsuite which indicate the required
features to run each test. So when we implement some feature, it will be
painless to reenable all its tests. In addition failing tests now have proper

> tickets in GHC's gitlab.

Reviewer 5: ```suggestion
tickets in GHC's GitLab.
```
Or just issue tracker

We've spent some time trying to run the testsuite on CI. Sadly Hadrian doesn't
support this yet (more concretely, it doesn't properly support running the

> testsuite for a cross-compiler in a bindist specified with `--test-compiler`).

Author: I've removed the technical details as most readers won't care.

out there, we replaced a FastString with its FastString unique, so it was only
made possible after removing Text and switching from Text to FastString.

#### Removal of custom file extensions and support for JS pragmas

GHCJS used the `.js.pp` file extension to identify JS files that needed to be
passed through CPP before being valid JavaScript. Adding support for this
extension in both Hadrian and GHC proved to be more painful than just adding
support for JS pragmas. So we decided to do the latter; similarly to Haskell
extension pragmas, you can now write "//#OPTIONS: CPP" in your JS files to
enable the CPP pass, and the file extension is always `.js`.

While we're on the topic of file extensions, technically `.js` files don't have
to be compiled into `.o` files (contrary to C/C++/Haskell/etc. files) at all.
However, build systems (Hadrian, Cabal...) and compilers (GHC) expect this. So
for consistency with other backends, we've added a "compilation" pass for `.js`
files too. They are now renamed into `.o` files with a "//JAVASCRIPT" header added
to distinguish them from object files produced by the JS backend (and from
Emscripten, in the future).

> #### So much cleanup and documentation

Reviewer 4: but it was so much!!
(I agree with the suggestion though)

([#20964](https://gitlab.haskell.org/ghc/ghc/-/issues/20964)/[!7377](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7377)).
This method doesn't require GHC to load module interfaces for the plugin and its
dependencies, hence workarounds GHC's limitations.

## What about libraries using C sources?

Libraries that use C sources (`c-sources` Cabal stanza) aren't supported by the
JS backend.
In the future we plan to use Emscripten to compile C sources and then to
generate some adapter code for them, but this isn't done yet.

For now there are two ways to fix libraries that use C sources.
The C code can either be rewritten in Javascript, or it can be rewritten in
Haskell.
Then it is possible to use Cabal predicates (e.g. `arch(js)`) to select between
the different versions.

We do have a preference for writing pure Haskell versions because it is more
future proof.
For example if someone adds some new backends for Lua, Java, CLR, etc. then the
Haskell version can be directly compiled by that backend and there is no extra work.
In contrast, if the C source is rewritten in JavaScript, then it would need to
be rewritten _for each_ backend.

> That is the approach we've taken when we wrote the ghc-bignum library.

Reviewer 5: Consider backticks around identifiers

provide shims so these JS sources will have to be upstreamed or reimplemented in
Haskell.

Note that the linking behavior is different due to the interpreted nature of
Javascript. In the JS backend, we can link with libraries using foreign imports
_even if_ the imported functions don't exist. Instead of failing at link time
(what usually happens with native code) a JS exception is raised only when and
if the imported function is called.

## How to help?

We have now reached our first milestone; anyone can easily build and test the JS
backend, and anyone can open bug reports or offer patches for the JS backend on
GHC's gitlab.

For those who offered their help this year: thank you! Until now it was
difficult to split the work into independent tasks (one fix led to a new
failure, which led to an architectural issue, etc.) and it was difficult to
coordinate with people outside of our team. However, we're now in a much better
position to discuss suggestions and to test/review patches in the spirit of open
source.

## tl;dr Just tell me how to say hello world {#build}

You need:

  - [emscripten](https://emscripten.org/docs/getting_started/downloads.html) version 3.14 or better

>   - llvm 15, or a patched llvm 

Reviewer 5: ```suggestion
  - LLVM 15, or a patched LLVM 
```

>   - nodejs, latest stable version but older should also work and only if you want to run the JS with node.

Reviewer 5: I can't parse this sentence.

Reviewer 2: resolved in 6c890a881dea5ecd1d93e9925b319563f4914a1a

   xelatex      : /nix/store/8jc2258h4nqzqjy303zzkssd3ip675pf-texlive-combined-2021/bin/xelatex
   makeinfo     : /run/current-system/sw/bin/makeinfo
   git          : /nix/store/vsr2cn15h7cbwd5vqsam2ab2jzwfbyf9-git-2.36.0/bin/git
   cabal-install : /nix/store/cjmd2qv1b5pdw4lxh1aw4xwwy4ibnb2p-cabal-install-3.6.2.0/bin/cabal

   Using LLVM tools
      clang : clang
      llc   : llc
      opt   : opt

   HsColour was not found; documentation will not contain source links

   Tools to build Sphinx HTML documentation available: YES
   Tools to build Sphinx PDF documentation available: YES
   Tools to build Sphinx INFO documentation available: YES
----------------------------------------------------------------------
```

Be sure to verify that `ar`, `ld`, `nm` and friends point to the emscripten
versions, i.e., the output shows `<tool> : <some-path>-emscripten-<tool>`.

#### Build the JS backend
```
./hadrian/build --bignum=native -j
```

#### Now compile hello world
```hs
module Main where

main :: IO ()
main = putStrLn "Hello JS!"
```

```
$ <path-to-ghc-root-dir>/_build/ghc-stage1 -fforce-recomp Main.hs

> $ ./Main

Reviewer 5: It would be nice if there was a bit of text saying what actually happens here. Is `Main` a script that launches node in the background?

Reviewer 2: resolved in dacd6320fa777e1f7039399b86986368da8dbc53
